File: backend/app/schemas/procurement_report.py
"""
Pydantic Schemas for Part 8E Procurement Reports (Tender Summary & Bid Evaluation)
"""

import uuid
from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, ConfigDict, Field


class ReportTenderInfo(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    tender_id: uuid.UUID
    tender_number: str
    title: str
    status: str
    organization_name: str
    category: Optional[str] = None
    procurement_type: Optional[str] = None
    currency: str = "INR"
    estimated_value: Optional[float] = None
    published_at: Optional[datetime] = None
    submission_end_date: Optional[datetime] = None


class ReportBidderInfo(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    organization_id: uuid.UUID
    name: str
    pan_number: Optional[str] = None
    gstin: Optional[str] = None
    udyam_number: Optional[str] = None
    business_type: Optional[str] = None
    state: Optional[str] = None
    city: Optional[str] = None


class ReportBidInfo(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    bid_id: uuid.UUID
    bid_number: str
    status: str
    submitted_at: Optional[datetime] = None
    quoted_amount: Optional[float] = None
    currency: str = "INR"
    is_shortlisted: bool = False
    shortlist_reason: Optional[str] = None


class ReportComplianceSection(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    evaluation_complete: bool = False
    evaluation_version: int = 1
    total_requirements: int = 0
    passed_count: int = 0
    failed_count: int = 0
    review_count: int = 0
    pending_count: int = 0
    not_applicable_count: int = 0
    mandatory_failures_count: int = 0
    critical_failures_count: int = 0


class ReportScoreSection(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    overall_compliance_score: Optional[float] = None
    score_type: str = "FINAL"
    scoring_complete: bool = False
    earned_weight: float = 0.0
    eligible_weight: float = 0.0
    category_scores: Dict[str, Any] = Field(default_factory=dict)
    scoring_version: int = 1
    is_stale: bool = False


class ReportRiskSection(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    base_risk_score: Optional[float] = None
    base_risk_level: Optional[str] = None
    adjusted_risk_score: Optional[float] = None
    adjusted_risk_level: Optional[str] = None
    override_applied: bool = False
    applied_overrides: List[Dict[str, Any]] = Field(default_factory=list)
    risk_complete: bool = False
    risk_version: int = 1
    is_stale: bool = False


class ReportDefectItem(BaseModel):
    requirement_code: str
    requirement_name: str
    category: str
    compliance_status: str
    is_mandatory: bool
    is_critical: bool
    reason: Optional[str] = None


class ReportHumanReviewItem(BaseModel):
    id: uuid.UUID
    review_type: str
    severity: str
    status: str
    resolution: Optional[str] = None
    reason: Optional[str] = None
    resolved_by_name: Optional[str] = None
    resolved_at: Optional[datetime] = None
    notes_count: int = 0


class ReportAISection(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    recommendation: Optional[str] = None
    recommendation_reason: Optional[str] = None
    summary: Optional[str] = None
    strengths: List[str] = Field(default_factory=list)
    concerns: List[str] = Field(default_factory=list)
    model_provider: Optional[str] = None
    model_name: Optional[str] = None
    prompt_version: Optional[str] = None
    guardrail_applied: bool = False
    guardrail_reason: Optional[str] = None
    confidence_label: Optional[str] = None
    is_stale: bool = False
    advisory_disclaimer: str = (
        "AI evaluation recommendation is strictly advisory assistance grounded in deterministic facts. "
        "AI does not make qualification, selection, or tender award decisions."
    )


class ReportFinalDecisionSection(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    decision: str = "NOT_DECIDED"
    reason: Optional[str] = None
    decision_summary: Optional[str] = None
    category: Optional[str] = None
    decided_by_name: Optional[str] = None
    decided_by_role: Optional[str] = None
    decided_at: Optional[datetime] = None
    decision_version: int = 0
    is_current: bool = True
    is_stale: bool = False
    stale_reason: Optional[str] = None


class ReportDecisionHistoryItem(BaseModel):
    decision_version: int
    decision: str
    reason: str
    decision_summary: Optional[str] = None
    decided_by_name: Optional[str] = None
    decided_at: datetime
    is_current: bool
    superseded_at: Optional[datetime] = None


class ReportAuditEventSummaryItem(BaseModel):
    event_type: str
    event_label: str
    action: str
    actor_name: str
    actor_source: str
    summary: str
    created_at: datetime


class BidEvaluationReportResponse(BaseModel):
    """
    Comprehensive, auditable Bid Compliance & Decision Dossier.
    """
    model_config = ConfigDict(from_attributes=True)

    report_id: uuid.UUID = Field(default_factory=uuid.uuid4)
    report_title: str = "Bid Compliance Verification & Evaluation Report"
    generated_at: datetime
    generated_by: str

    tender: ReportTenderInfo
    bidder: ReportBidderInfo
    bid: ReportBidInfo

    compliance: ReportComplianceSection
    score: ReportScoreSection
    risk: ReportRiskSection

    mandatory_failures: List[ReportDefectItem] = Field(default_factory=list)
    critical_findings: List[ReportDefectItem] = Field(default_factory=list)
    human_reviews: List[ReportHumanReviewItem] = Field(default_factory=list)

    ai_recommendation: Optional[ReportAISection] = None
    final_human_decision: ReportFinalDecisionSection
    decision_history: List[ReportDecisionHistoryItem] = Field(default_factory=list)

    stale_warnings: List[str] = Field(default_factory=list)
    mock_verification_disclaimer: Optional[str] = None
    audit_timeline: List[ReportAuditEventSummaryItem] = Field(default_factory=list)


class TenderSummaryBidItem(BaseModel):
    bid_id: uuid.UUID
    bid_number: str
    bidder_name: str
    quoted_amount: Optional[float] = None
    compliance_score: Optional[float] = None
    adjusted_risk_level: Optional[str] = None
    human_decision_status: str
    is_shortlisted: bool = False
    critical_defects_count: int = 0
    open_reviews_count: int = 0


class TenderReportResponse(BaseModel):
    """
    Comprehensive Tender Evaluation Summary Report for Procurement Officers & Review Committees.
    """
    model_config = ConfigDict(from_attributes=True)

    report_id: uuid.UUID = Field(default_factory=uuid.uuid4)
    report_title: str = "Tender Compliance Evaluation Summary Report"
    generated_at: datetime
    generated_by: str

    tender: ReportTenderInfo

    # Key Aggregates
    total_bids_submitted: int = 0
    total_bids_evaluated: int = 0
    total_qualified: int = 0
    total_disqualified: int = 0
    total_under_review: int = 0
    total_not_decided: int = 0
    total_shortlisted: int = 0

    # Risk Distribution
    risk_distribution: Dict[str, int] = Field(
        default_factory=lambda: {"LOW": 0, "MEDIUM": 0, "HIGH": 0, "CRITICAL": 0}
    )

    # Compliance Overview
    average_compliance_score: Optional[float] = None
    total_critical_defects: int = 0
    total_open_reviews: int = 0

    # Itemized Bids
    bids: List[TenderSummaryBidItem] = Field(default_factory=list)

File: frontend/src/types/procurement_report.ts
export interface ReportTenderInfo {
  tender_id: string;
  tender_number: string;
  title: string;
  status: string;
  organization_name: string;
  category?: string | null;
  procurement_type?: string | null;
  currency: string;
  estimated_value?: number | null;
  published_at?: string | null;
  submission_end_date?: string | null;
}

export interface ReportBidderInfo {
  organization_id: string;
  name: string;
  pan_number?: string | null;
  gstin?: string | null;
  udyam_number?: string | null;
  business_type?: string | null;
  state?: string | null;
  city?: string | null;
}

export interface ReportBidInfo {
  bid_id: string;
  bid_number: string;
  status: string;
  submitted_at?: string | null;
  quoted_amount?: number | null;
  currency: string;
  is_shortlisted: boolean;
  shortlist_reason?: string | null;
}

export interface ReportComplianceSection {
  evaluation_complete: boolean;
  evaluation_version: number;
  total_requirements: number;
  passed_count: number;
  failed_count: number;
  review_count: number;
  pending_count: number;
  not_applicable_count: number;
  mandatory_failures_count: number;
  critical_failures_count: number;
}

export interface ReportScoreSection {
  overall_compliance_score?: number | null;
  score_type: string;
  scoring_complete: boolean;
  earned_weight: number;
  eligible_weight: number;
  category_scores: Record<string, any>;
  scoring_version: number;
  is_stale: boolean;
}

export interface ReportRiskSection {
  base_risk_score?: number | null;
  base_risk_level?: string | null;
  adjusted_risk_score?: number | null;
  adjusted_risk_level?: string | null;
  override_applied: boolean;
  applied_overrides: Array<Record<string, any>>;
  risk_complete: boolean;
  risk_version: number;
  is_stale: boolean;
}

export interface ReportDefectItem {
  requirement_code: string;
  requirement_name: string;
  category: string;
  compliance_status: string;
  is_mandatory: boolean;
  is_critical: boolean;
  reason?: string | null;
}

export interface ReportHumanReviewItem {
  id: string;
  review_type: string;
  severity: string;
  status: string;
  resolution?: string | null;
  reason?: string | null;
  resolved_by_name?: string | null;
  resolved_at?: string | null;
  notes_count: number;
}

export interface ReportAISection {
  recommendation?: string | null;
  recommendation_reason?: string | null;
  summary?: string | null;
  strengths: string[];
  concerns: string[];
  model_provider?: string | null;
  model_name?: string | null;
  prompt_version?: string | null;
  guardrail_applied: boolean;
  guardrail_reason?: string | null;
  confidence_label?: string | null;
  is_stale: boolean;
  advisory_disclaimer: string;
}

export interface ReportFinalDecisionSection {
  decision: string;
  reason?: string | null;
  decision_summary?: string | null;
  category?: string | null;
  decided_by_name?: string | null;
  decided_by_role?: string | null;
  decided_at?: string | null;
  decision_version: number;
  is_current: boolean;
  is_stale: boolean;
  stale_reason?: string | null;
}

export interface ReportDecisionHistoryItem {
  decision_version: number;
  decision: string;
  reason: string;
  decision_summary?: string | null;
  decided_by_name?: string | null;
  decided_at: string;
  is_current: boolean;
  superseded_at?: string | null;
}

export interface ReportAuditEventSummaryItem {
  event_type: string;
  event_label: string;
  action: string;
  actor_name: string;
  actor_source: string;
  summary: string;
  created_at: string;
}

export interface BidEvaluationReportResponse {
  report_id: string;
  report_title: string;
  generated_at: string;
  generated_by: string;
  tender: ReportTenderInfo;
  bidder: ReportBidderInfo;
  bid: ReportBidInfo;
  compliance: ReportComplianceSection;
  score: ReportScoreSection;
  risk: ReportRiskSection;
  mandatory_failures: ReportDefectItem[];
  critical_findings: ReportDefectItem[];
  human_reviews: ReportHumanReviewItem[];
  ai_recommendation?: ReportAISection | null;
  final_human_decision: ReportFinalDecisionSection;
  decision_history: ReportDecisionHistoryItem[];
  stale_warnings: string[];
  mock_verification_disclaimer?: string | null;
  audit_timeline: ReportAuditEventSummaryItem[];
}

export interface TenderSummaryBidItem {
  bid_id: string;
  bid_number: string;
  bidder_name: string;
  quoted_amount?: number | null;
  compliance_score?: number | null;
  adjusted_risk_level?: string | null;
  human_decision_status: string;
  is_shortlisted: boolean;
  critical_defects_count: number;
  open_reviews_count: number;
}

export interface TenderReportResponse {
  report_id: string;
  report_title: string;
  generated_at: string;
  generated_by: string;
  tender: ReportTenderInfo;
  total_bids_submitted: number;
  total_bids_evaluated: number;
  total_qualified: number;
  total_disqualified: number;
  total_under_review: number;
  total_not_decided: number;
  total_shortlisted: number;
  risk_distribution: Record<string, number>;
  average_compliance_score?: number | null;
  total_critical_defects: number;
  total_open_reviews: number;
  bids: TenderSummaryBidItem[];
}
